base/string.jl

```diff
@@ -968,8 +968,8 @@ byte_string_classify(s::ByteString) = byte_string_classify(s.data)
     # 1: valid ASCII
     # 2: valid UTF-8
 
-is_valid_ascii(s::Union(Array{UInt8,1},ByteString)) = byte_string_classify(s) == 1
-is_valid_utf8(s::Union(Array{UInt8,1},ByteString)) = byte_string_classify(s) != 0
+isvalid(::Type{ASCIIString}, s::Union(Array{UInt8,1},ByteString)) = byte_string_classify(s) == 1
```

Comment:
Regarding the technical note: yes.  `utf16.jl` currently defines `convert(::Type{UTF16String}, s::AbstractString)`.   If you define a more specialized/optimized conversion routine `Base.convert(::Type{UTF16String}, s::UTF8String)`, for example, and a function in Base subsequently calls e.g. `utf16(....some UTF8String....)`, then it will dispatch to your method.    The exception would be if that function in Base has already compiled and inlined the old conversion call prior to your new definition.   This might have happened, for example, on Windows for a Win32 filesystem call that was already made while booting Julia, but of course the argument string conversion is negligible anyway for a filesystem call.

Adding new methods for more specific argument signatures is not the same thing as "monkey patching" in the Python sense.

I wish you would stop with the comments to the effect that you are the only one who cares about string-handling performance.   As I said in #11004, everyone cares, but we also realize that there are tradeoffs to code bloat for the sake of performance.